### ISSUE TITLE
Postgresql ARM provider Bug

### ISSUE THREAD

**Issue by lawrencegripper** [Azure/azure-sdk-for-go#728]
Hi, 

I believe there is a bug in the Postgresql ARM provider here:

https://github.com/Azure/azure-sdk-for-go/blob/master/arm/postgresql/models.go#L227

When calling the CreateOrUpdate method I pass in a [ServerForCreate](https://github.com/Azure/azure-sdk-for-go/blob/master/arm/postgresql/models.go#L227) struct. This struct has a "Properties" which is set to an interface of [ServerPropertiesForCreate](https://github.com/Azure/azure-sdk-for-go/blob/master/arm/postgresql/models.go#L320). 

The validation initially failed as I didn't pass in a pointer of "ServerPropertiesForDefaultCreate", the error returned stated "parameters.Properties" does not exist which threw me for a bit!

Passing a pointer allows the code to proceed to submitting the request but it doesn't provision a server. I believe autorest incorrectly serializes the type and the "Properties" aren't submitted with the request. The behavior I see is the request taking a large amount of time then timing out. 

![screen shot 2017-08-17 at 15 14 01](https://user-images.githubusercontent.com/1939288/29416745-fda9c220-835e-11e7-9e62-ce6ce4a7bb81.png)

In my code I've worked around the issue by replacing the interface "ServicePropertiesForCreate" with the type [ServerPropertiesForDefaultCreate](https://github.com/lawrencegripper/kube-azcon/blob/master/vendor/github.com/Azure/azure-sdk-for-go/arm/postgresql/models.go#L227) and this succeeds. 

You can see a full example of my usage [of the sdk here](https://github.com/lawrencegripper/kube-azcon/blob/master/azureProviders/postrgres.go).  

I'm relatively new to Golang so apologies if this is down to something I've misunderstood on my end!

**Comment by mcardosos** [comment]
Hello @lawrencegripper !
The SDK is meant to be used just as you described. :) The `Properties` field inside the `ServerForCreate` struct is a `ServerPropertiesForCreate` interface, which is satisfied by `ServerPropertiesForDefaultCreate` (`ServerPropertiesForRestore` also satisfies this interface). Therefore you can use either `ServerPropertiesForDefaultCreate` or `ServerPropertiesForRestore` structs as `Properties`. [Cool interfaces example](https://gobyexample.com/interfaces).

I was able to create a server. My code looks like this...
``` go
server := postgresql.ServerForCreate{
	Location: to.StringPtr("westus"),
	Properties: &postgresql.ServerPropertiesForDefaultCreate{
		AdministratorLogin:         to.StringPtr("notadmin"),
		AdministratorLoginPassword: to.StringPtr("Pa$$w0rd1975"),
		StorageMB:                  to.Int64Ptr(51200),
	},
}
_, errChan := client.CreateOrUpdate("resourcegroup", "servername", server, nil)
```
and the request body looks like this...
```json
{"properties":{"createMode":"Default","storageMB":51200,"administratorLogin":"notadmin","administratorLoginPassword":"Pa$$w0rd1975"},"location":"westus"}
```
The SDK is working correctly here.

As for the request taking lots of time and then failing, I will investigate more.

**Comment by mcardosos** [comment]
For the resource operation failure, it is something that can happen (it is an error server side). The thing to to on the client side is retry the operation.

**Comment by lawrencegripper** [comment]
Thanks, instrumented the code a little more and you dead right. Sorry!

I believe the issue I was seeing with long deployment time & failures was due to some code I wrote to generate passwords, some of which appear to cause issues with the postgres ARM provider. It looks like, while I was testing yesterda,y this random chance of failure fooled me into thinking it was an SDK issue - as it deployed correctly by chance after I'd tweaked the SDK 🤦‍♂️
